code/datums/components/storage.dm

```diff
@@ -0,0 +1,335 @@
+
+/datum/component/storage
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS // old component gets passed new component's initialization args
+	var/datum/hud/storage/hud
+	// future: add stored_contents list and use that to keep track instead of source.contents
+	var/list/can_hold // if this exists, the storage checks it for what can be held
+	var/in_list_or_max = FALSE // if this is true, the storage allows items that are either in its can_hold list or items that have a lesser w_class than max_wclass
+	var/max_wclass = 2
+	var/slots = 7
+	var/sneaky = FALSE // does it play a sound when used?
+	var/does_not_open_in_pocket = TRUE
+
+/datum/component/storage/Initialize(var/list/spawn_contents, var/list/can_hold, var/in_list_or_max = FALSE, var/max_wclass = 2, var/slots = 7, var/sneaky = FALSE, var/does_not_open_in_pocket = TRUE)
+	if (!isobj(parent)) // future: allow all movable atoms
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, list(COMSIG_ATOM_ATTACK_BY), .proc/attack_by)
+	RegisterSignal(parent, list(COMSIG_ATOM_HAND_ATTACK), .proc/hand_attack)
+	RegisterSignal(parent, list(COMSIG_ATOM_MOUSE_DROP), .proc/mouse_drop)
+	RegisterSignal(parent, list(COMSIG_OBJ_MOVE_TRIGGER), .proc/move_trigger)
+	RegisterSignal(parent, list(COMSIG_ITEM_ATTACK_SELF), .proc/attack_self)
+	RegisterSignal(parent, list(COMSIG_ITEM_ATTACK_POST), .proc/attack)
+	RegisterSignal(parent, list(COMSIG_ITEM_AFTER_ATTACK), .proc/after_attack)
+	RegisterSignal(parent, list(COMSIG_ITEM_DROPPED), .proc/on_dropped)
+	RegisterSignal(parent, list(COMSIG_ITEM_BUILD_TOOLTIP), .proc/build_tooltip)
+	RegisterSignal(parent, list(COMSIG_STORAGE_GET_CONTENTS), .proc/get_contents)
+	RegisterSignal(parent, list(COMSIG_STORAGE_GET_ALL_CONTENTS), .proc/get_all_contents)
+	RegisterSignal(parent, list(COMSIG_STORAGE_FIND_TYPE), .proc/find_type)
+	RegisterSignal(parent, list(COMSIG_STORAGE_TRANSFER_ITEM), .proc/transfer_item)
+	RegisterSignal(parent, list(COMSIG_STORAGE_CAN_FIT), .proc/can_fit)
+	RegisterSignal(parent, list(COMSIG_MOVABLE_EMP_ACT), .proc/emp_act)
+	hud = new(src)
+	src.can_hold = can_hold
+	src.in_list_or_max = in_list_or_max
+	src.max_wclass = max_wclass
+	src.slots = slots
+	src.sneaky = sneaky
+	src.does_not_open_in_pocket = does_not_open_in_pocket
+	if (spawn_contents)
+		make_my_stuff(spawn_contents)
+
+// since we have COMPONENT_DUPE_UNIQUE_PASSARGS, which means the old component gets passed the new component's initialization args, we should always get SC = null, original = TRUE (since new component doesn't get passed)
+// we should only get passed the other args (the new component's initialization args), if applicable
+/datum/component/storage/InheritComponent(datum/component/storage/SC, original, var/list/spawn_contents, var/list/can_hold, var/in_list_or_max, var/max_wclass, var/slots, var/sneaky, var/does_not_open_in_pocket)
+	if (spawn_contents)
+		make_my_stuff(spawn_contents)
+	if (can_hold)
+		src.can_hold = can_hold
+	if (!isnull(in_list_or_max))
```

Comment:
The `!isnull()` checks look kinda ugly but I don't think there's a better way?

Reply:
That's fine and imo it's a pretty okay pattern to use with arguments.